c++/src/kj/async-inl.h

```diff
@@ -300,6 +300,77 @@ class PromiseNode: private AsyncObject {
 
 // -------------------------------------------------------------------
 
+class Environment: public kj::Refcounted {
+  // Base class of Environment::Holder<T>. This class also houses and provides access to the
+  // thread_local reference to the currently-active Environment.
+
+public:
+  class Scope;
+  template <typename T> class Holder;
+
+  KJ_DISALLOW_COPY(Environment);
+  ~Environment() noexcept(false) = 0;
+
+  static Maybe<Own<Environment>> tryCaptureCurrent();
+  static Maybe<Environment&> tryGetCurrent();
+
+private:
+  Environment() = default;
+
+  static thread_local Maybe<Environment&> current;
```

Comment:
Two problems with this `thread_local`:
* Defining it in a header accessible from multiple translation units forces the compiler to add extra boilerplate when accessing it. Defining it as a global with static linkage in the `.c++` file reduces this boilerplate, and since you are only actually accessing it from inside the `.c++` file anyway, there's no downside.
* Defining it as `Maybe<Environment&>` may make the compiler believe this is a non-trivial value with construction and destruction requirements. I'm not sure if it's smart enough to see that the constructor and destructor are actually trivial. We definitely don't want the compiler to believe it has to arrange callbacks for thread start/stop. I would define it as `Environment*` instead to make things obvious for the compiler.

Reply:
Actually does this `thread_local` need to exist at all or can you store the information in the `EventLoop` which already has a `thread_local` pointing to it?

Reply:
Hmm, detached promises are actually owned by the EventLoop. However, the EventLoop is only "current" for the thread during the time when the `WaitScope` exists, which is destroyed before the `EventLoop` is destroyed, so possibly before any detached promises are destroyed.

There's actually a `WaitScope::cancelAllDetached()` which exists specifically to work around this, destroying detached promises while the `EventLoop` is still current.

But why does this matter for environments exactly? Why does the promise's destructor cane if the event loop is still current?

Reply:
> But why does this matter for environments exactly? Why does the promise's destructor cane if the event loop is still current?

It determines whether we can provide the following guarantee: "All user code scheduled asynchronously by `f()` from within `kj::runInEnvironment(E(e), f)` will have access to `e` by calling `kj::getEnvironment<E>()`." The "user code scheduled asynchronously" that I'm talking about includes things like `.then()` continuation destructors, `.attach()` attachment destructors, coroutine frame object destructors, etc.

If we don't require the event loop to be current, then we must accept that we also don't have access to the current environment ref (the line of code we're commenting on, now moved to the EventLoop). I think this means that `kj::getEnvironment<E>()` must throw if there's no current event loop, breaking that guarantee.

Reply:
I made an attempt at moving the ref to the EventLoop in the latest three commits. The first one does the move, the second adds `cancelAllDetached()` to `~WaitScope()` to fix the issues with detached promises. The third fixes some bugs in my Fiber changes that were surfaced by the move to the EventLoop.

Regarding `cancelAllDetached()`, that was my first hacky impulse, but perhaps instead we should make a WaitScope in EventLoop's destructor for the purpose of dropping the daemons?

Reply:
> perhaps instead we should make a WaitScope in EventLoop's destructor for the purpose of dropping the daemons?

I did this.